autopilot/hostscore.go

```diff
@@ -0,0 +1,256 @@
+package autopilot
+
+import (
+	"math"
+	"math/big"
+	"sort"
+	"time"
+
+	"go.sia.tech/renterd/hostdb"
+	"go.sia.tech/siad/build"
+	"go.sia.tech/siad/types"
+	"lukechampine.com/frand"
+)
+
+const (
+	// maxBaseRPCPriceVsBandwidth is the max ratio for sane pricing between the
+	// MinBaseRPCPrice and the MinDownloadBandwidthPrice. This ensures that 1
+	// million base RPC charges are at most 1% of the cost to download 4TB. This
+	// ratio should be used by checking that the MinBaseRPCPrice is less than or
+	// equal to the MinDownloadBandwidthPrice multiplied by this constant
+	maxBaseRPCPriceVsBandwidth = uint64(40e3)
+
+	// maxSectorAccessPriceVsBandwidth is the max ratio for sane pricing between
+	// the MinSectorAccessPrice and the MinDownloadBandwidthPrice. This ensures
+	// that 1 million base accesses are at most 10% of the cost to download 4TB.
+	// This ratio should be used by checking that the MinSectorAccessPrice is
+	// less than or equal to the MinDownloadBandwidthPrice multiplied by this
+	// constant
+	maxSectorAccessPriceVsBandwidth = uint64(400e3)
+)
+
+type hostScore struct {
+	host  host
+	score float64
+}
+
+func newHostScore(h host) *hostScore {
+	return &hostScore{
+		host:  h,
+		score: 1,
+	}
+}
+
+func (s *hostScore) finalize() float64 {
+	return s.score
+}
+
+func (s *hostScore) withAgeScore() *hostScore {
+	const day = 24 * time.Hour
+	weights := []struct {
+		age    time.Duration
+		factor float64
+	}{
+		{128 * day, 1.5},
+		{64 * day, 2},
+		{32 * day, 2},
+		{16 * day, 2},
+		{8 * day, 3},
+		{4 * day, 3},
+		{2 * day, 3},
+		{1 * day, 3},
+	}
+
+	age := time.Since(s.host.Announcements[0].Timestamp)
+	weight := 1.0
+	for _, w := range weights {
+		if age >= w.age {
+			break
+		}
+		weight /= w.factor
+	}
+
+	s.score *= weight
```

Comment:
(nit) could save a line here with a helper like `return s.addWeight(weight)`. Which makes it a bit more obvious that the weight was applied to the running total.

Reply:
added in #36